### PR TITLE
Add and remove self loops

### PR DESCRIPTION
**Pull request by francescofarina**
## Proposed changes

Functions to add and remove self loops (https://github.com/TristanBilot/mlx-graphs/issues/38)

## Checklist

Put an `x` in the boxes that apply.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation

### CODE REVIEW

**Comment by francescofarina** [comment]
The function `get_unique_edge_indices` is probably not super efficient, but it seems there's not much more one can do right now. I may be missing something though, so feel free to suggest another solution.

File: tests/utils/test_transformations.py [hunk 1]
```diff
@@ -104,3 +107,87 @@ def test_to_adjacency_matrix():
...
+    expected_y = mx.concatenate([edge_features, mx.ones([3, 2]) * fill], 0)
+    assert mx.array_equal(y, expected_y)
+    assert mx.array_equal(x, expected_x)
+
+    # don't allow repeated
+    x, y = add_self_loops(edge_index, edge_features, allow_repeated=False)
+    expected_x = mx.array([[0, 0, 1, 1, 1, 2], [0, 1, 0, 2, 1, 2]])
+    expected_y = mx.ones([6, 2])
+    print(x)
+    print(expected_x)
```
Comment (TristanBilot) [inline tests/utils/test_transformations.py:167]: missed those ;)

File: mlx_graphs/utils/transformations.py [hunk 2]
```diff
@@ -155,3 +155,116 @@ def get_src_dst_features(
...
+                                    [2, 1, 2],
+                                ])
+        x = remove_common_edges(edge_index_1, edge_index_2)
+        # [0, 1]
+    """
+    edge_1_tuples = [tuple(edge.tolist()) for edge in edge_index_1.transpose()]
+    edge_2_tuples = [tuple(edge.tolist()) for edge in edge_index_2.transpose()]
+
+    return mx.array(
+        [i for i, edge in enumerate(edge_1_tuples) if edge not in edge_2_tuples]
```
Comment (TristanBilot) [inline mlx_graphs/utils/transformations.py:191]: This loop in O(n^2) can be transformed in O(n) with a set:
```python
edge_1_tuples = [tuple(edge.tolist()) for edge in edge_index_1.transpose()]
edge_2_set = set(tuple(edge.tolist()) for edge in edge_index_2.transpose())

return mx.array(
    [i for i, edge in enumerate(edge_1_tuples) if edge not in edge_2_set]
)
```

File: mlx_graphs/utils/transformations.py [hunk 3]
```diff
@@ -155,3 +155,116 @@ def get_src_dst_features(
...
+    .. code-block:: python
+
+        edge_index_1 = mx.array([
+                                    [0, 1, 1, 2],
+                                    [1, 0, 2, 1],
+                                ])
+        edge_index_2 = mx.array([
+                                    [1, 2, 2],
+                                    [2, 1, 2],
+                                ])
```
Comment (TristanBilot) [inline mlx_graphs/utils/transformations.py:183]: Too many indentations on the array values

File: mlx_graphs/utils/transformations.py [hunk 4]
```diff
@@ -155,3 +155,116 @@ def get_src_dst_features(
...
+    return full_edge_index, full_edge_features
+
+
+@validate_edge_index_and_features
+def remove_self_loops(
+    edge_index: mx.array,
+    edge_features: Optional[mx.array] = None,
+) -> tuple[mx.array, Optional[mx.array]]:
+    """
+    Remove self-loops from the given graph represented by edge_index and edge_features.
```
Comment (TristanBilot) [inline mlx_graphs/utils/transformations.py:250]: Removes*

File: mlx_graphs/utils/transformations.py [hunk 5]
```diff
@@ -155,3 +155,116 @@ def get_src_dst_features(
         )
 
     return src_val, dst_val
+
+
+def get_unique_edge_indices(edge_index_1: mx.array, edge_index_2: mx.array) -> mx.array:
+    """
+    Compute the indices of the edges in edge_index_1 that are NOT present in edge_index_2
```
Comment (TristanBilot) [inline mlx_graphs/utils/transformations.py:162]: Computes*

File: mlx_graphs/utils/transformations.py [hunk 6]
```diff
@@ -155,3 +155,116 @@ def get_src_dst_features(
...
+    edge_features: Optional[mx.array] = None,
+    num_nodes: Optional[int] = None,
+    fill_value: Optional[Union[float, mx.array]] = 1,
+    allow_repeated: Optional[bool] = True,
+) -> tuple[mx.array, Optional[mx.array]]:
+    """
+    Adds self-loops to the given graph represented by edge_index and edge_features.
+
+    Args:
+        edge_index: the edge index of the graph, with shape [2, num_edges]
```
Comment (TristanBilot) [inline mlx_graphs/utils/transformations.py:207]: Edge index*?

File: mlx_graphs/utils/transformations.py [hunk 7]
```diff
@@ -155,3 +155,116 @@ def get_src_dst_features(
...
+@validate_edge_index_and_features
+def remove_self_loops(
+    edge_index: mx.array,
+    edge_features: Optional[mx.array] = None,
+) -> tuple[mx.array, Optional[mx.array]]:
+    """
+    Remove self-loops from the given graph represented by edge_index and edge_features.
+
+    Args:
+        edge_index: the edge index of the graph, with shape [2, num_edges]
```
Comment (TristanBilot) [inline mlx_graphs/utils/transformations.py:253]: Same, "Edge index"*?

**Review by TristanBilot — CHANGES_REQUESTED** [review]

**Comment by TristanBilot** [comment]
> The function `get_unique_edge_indices` is probably not super efficient, but it seems there's not much more one can do right now. I may be missing something though, so feel free to suggest another solution.

Check https://github.com/TristanBilot/mlx-graphs/pull/45#discussion_r1470156220

File: mlx_graphs/utils/transformations.py (hunk 2 above)
Comment (TristanBilot) [reply mlx_graphs/utils/transformations.py:191]: Unfortunately, I'm not aware of an optimal way to vectorize the for loop here even with numpy, but it should be ok even with some millions edges, you can try locally if you want to be sure.

Comment (francescofarina) [reply mlx_graphs/utils/transformations.py:191]: Yes, good catch!

**Comment by francescofarina** [comment]
Should be all good now.

**Review by TristanBilot — APPROVED** [review]
Ok for me!

**Pull request merged**